Simplify `slice::Iter::next` enough that it inlines

scottmcm opened the pull request:
Inspired by this zulip conversation: <https://rust-lang.zulipchat.com/#narrow/channel/189540-t-compiler.2Fwg-mir-opt/topic/Feedback.20on.20a.20MIR.20optimization.20idea/near/498579990>

~~Draft for now because it needs #136735 to get the codegen tests to pass.~~

scottmcm committed:
Simplify `slice::Iter::next` enough that it inlines

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit f7970b37cac3b383544ab1eb91f3cd2e7360dbe2 with merge 30df00cd8218095deb80cc6b913de02f5ae4a5b0...
<!-- homu: {"type":"TryBuildStarted","head_sha":"f7970b37cac3b383544ab1eb91f3cd2e7360dbe2","merge_sha":"30df00cd8218095deb80cc6b913de02f5ae4a5b0"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13224988732/job/36914732871) [(plain)](https://github.com/rust-lang/rust/commit/f7970b37cac3b383544ab1eb91f3cd2e7360dbe2/checks/36914732871/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#21 exporting to docker image format
#21 sending tarball 27.3s done
#21 DONE 33.7s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
failures:

---- [codegen] tests/codegen/issues/issue-37945.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-18/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-37945/issue-37945.ll" "/checkout/tests/codegen/issues/issue-37945.rs" "--check-prefix=CHECK" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
/checkout/tests/codegen/issues/issue-37945.rs:14:17: error: CHECK-NEXT: expected string not found in input
/checkout/tests/codegen/issues/issue-37945.rs:14:17: error: CHECK-NEXT: expected string not found in input
 // CHECK-NEXT: [[A:%.*]] = icmp ne ptr {{%xs.0|%xs.1}}, null
                ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-37945/issue-37945.ll:8:7: note: scanning from here
      ^
      ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-37945/issue-37945.ll:9:3: note: possible intended match here
 %_3.not = icmp eq ptr %xs.1, %xs.0
/checkout/tests/codegen/issues/issue-37945.rs:27:17: error: CHECK-NEXT: expected string not found in input
/checkout/tests/codegen/issues/issue-37945.rs:27:17: error: CHECK-NEXT: expected string not found in input
 // CHECK-NEXT: [[C:%.*]] = icmp ne ptr {{%xs.0|%xs.1}}, null
                ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-37945/issue-37945.ll:15:7: note: scanning from here
      ^
      ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-37945/issue-37945.ll:16:4: note: possible intended match here
 %_12.not = icmp eq ptr %xs.1, %xs.0

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-37945/issue-37945.ll
Check file: /checkout/tests/codegen/issues/issue-37945.rs


-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
           1: ; ModuleID = 'issue_37945.20c1cf91498ad042-cgu.0' 
           2: source_filename = "issue_37945.20c1cf91498ad042-cgu.0" 
           3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128" 
           5:  
           5:  
           6: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
           7: define noundef zeroext i1 @is_empty_1(ptr noundef readnone %xs.0, ptr noundef nonnull readnone %xs.1) unnamed_addr #0 { 
           8: start: 
next:14'0           X error: no match found
           9:  %_3.not = icmp eq ptr %xs.1, %xs.0 
next:14'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
next:14'1       ?                                  possible intended match
          10:  ret i1 %_3.not 
next:14'0     ~~~~~~~~~~~~~~~~
          11: } 
next:14'0     ~~
next:14'0     ~
next:14'0     ~
          13: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
next:14'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          14: define noundef zeroext i1 @is_empty_2(ptr noundef readnone %xs.0, ptr noundef nonnull readnone %xs.1) unnamed_addr #0 { 
next:14'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          15: start: 
next:27'0           X error: no match found
          16:  %_12.not = icmp eq ptr %xs.1, %xs.0 
next:27'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
next:27'1        ?                                  possible intended match
          17:  ret i1 %_12.not 
next:27'0     ~~~~~~~~~~~~~~~~~
          18: } 
next:27'0     ~~
next:27'0     ~
next:27'0     ~
          20: attributes #0 = { mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" } 
next:27'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
next:27'0     ~
next:27'0     ~
          22: !llvm.module.flags = !{!0, !1} 
next:27'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          23: !llvm.ident = !{!2} 
next:27'0     ~~~~~~~~~~~~~~~~~~~~
next:27'0     ~
next:27'0     ~
          25: !0 = !{i32 8, !"PIC Level", i32 2} 
next:27'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          26: !1 = !{i32 2, !"RtLibUseGOT", i32 1} 
next:27'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          27: !2 = !{!"rustc version 1.86.0-nightly (1e1da50a0 2025-02-09)"} 
next:27'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
------------------------------------------



```

</details>

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13225044900/job/36916702568)
Build commit: 30df00cd8218095deb80cc6b913de02f5ae4a5b0 (`30df00cd8218095deb80cc6b913de02f5ae4a5b0`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13225044900/job/36916702568"},"merge_sha":"30df00cd8218095deb80cc6b913de02f5ae4a5b0"} -->

rust-timer commented:
Queued 30df00cd8218095deb80cc6b913de02f5ae4a5b0 with parent 43ca9d18e333797f0aa3b525501a7cec8d61a96b, future [comparison URL](https://perf.rust-lang.org/compare.html?start=43ca9d18e333797f0aa3b525501a7cec8d61a96b&end=30df00cd8218095deb80cc6b913de02f5ae4a5b0).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.4 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (30df00cd8218095deb80cc6b913de02f5ae4a5b0): [comparison URL](https://perf.rust-lang.org/compare.html?start=43ca9d18e333797f0aa3b525501a7cec8d61a96b&end=30df00cd8218095deb80cc6b913de02f5ae4a5b0&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=43ca9d18e333797f0aa3b525501a7cec8d61a96b&end=30df00cd8218095deb80cc6b913de02f5ae4a5b0&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.4%  | [0.4%, 0.4%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 9.4%  | [9.4%, 9.4%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-2.2%, -0.1%] | 210   |
| Improvements ✅ <br /> (secondary) | -0.5% | [-1.1%, -0.1%] | 131   |
| All ❌✅ (primary)                 | -0.6% | [-2.2%, 0.4%]  | 211   |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=43ca9d18e333797f0aa3b525501a7cec8d61a96b&end=30df00cd8218095deb80cc6b913de02f5ae4a5b0&stat=max-rss)
<details>
<summary>Results (primary -2.2%, secondary -2.5%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.4%  | [1.7%, 2.8%]   | 3     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -4.4% | [-8.5%, -2.4%] | 6     |
| Improvements ✅ <br /> (secondary) | -2.5% | [-2.5%, -2.5%] | 1     |
| All ❌✅ (primary)                 | -2.2% | [-8.5%, 2.8%]  | 9     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=43ca9d18e333797f0aa3b525501a7cec8d61a96b&end=30df00cd8218095deb80cc6b913de02f5ae4a5b0&stat=cycles:u)
<details>
<summary>Results (primary -1.0%, secondary 1.7%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 4.6%  | [4.6%, 4.6%]   | 1     |
| Improvements ✅ <br /> (primary)   | -1.0% | [-1.1%, -0.9%] | 2     |
| Improvements ✅ <br /> (secondary) | -1.1% | [-1.1%, -1.1%] | 1     |
| All ❌✅ (primary)                 | -1.0% | [-1.1%, -0.9%] | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=43ca9d18e333797f0aa3b525501a7cec8d61a96b&end=30df00cd8218095deb80cc6b913de02f5ae4a5b0&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.1%, secondary -0.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5%  | [0.0%, 5.0%]   | 28    |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.2% | [-0.5%, -0.0%] | 45    |
| Improvements ✅ <br /> (secondary) | -0.1% | [-1.3%, -0.0%] | 48    |
| All ❌✅ (primary)                 | 0.1%  | [-0.5%, 5.0%]  | 73    |
</details>

**Bootstrap**: 780.488s -> 778.559s (-0.25%)
**Artifact size**: 329.04 MiB -> 329.26 MiB (0.06%)

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13260228104/job/37021396819)
Build commit: eaf73cd9c4fc608d616056b232ccb9cbb8df5679 (`eaf73cd9c4fc608d616056b232ccb9cbb8df5679`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13260228104/job/37021396819"},"merge_sha":"eaf73cd9c4fc608d616056b232ccb9cbb8df5679"} -->

rust-timer commented:
Queued eaf73cd9c4fc608d616056b232ccb9cbb8df5679 with parent c182ce9cbc8c29ebc1b4559d027df545e6cdd287, future [comparison URL](https://perf.rust-lang.org/compare.html?start=c182ce9cbc8c29ebc1b4559d027df545e6cdd287&end=eaf73cd9c4fc608d616056b232ccb9cbb8df5679).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (eaf73cd9c4fc608d616056b232ccb9cbb8df5679): [comparison URL](https://perf.rust-lang.org/compare.html?start=c182ce9cbc8c29ebc1b4559d027df545e6cdd287&end=eaf73cd9c4fc608d616056b232ccb9cbb8df5679&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=c182ce9cbc8c29ebc1b4559d027df545e6cdd287&end=eaf73cd9c4fc608d616056b232ccb9cbb8df5679&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.4%  | [0.4%, 0.5%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 9.3%  | [9.3%, 9.3%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-2.2%, -0.1%] | 196   |
| Improvements ✅ <br /> (secondary) | -0.5% | [-2.4%, -0.1%] | 101   |
| All ❌✅ (primary)                 | -0.6% | [-2.2%, 0.5%]  | 198   |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=c182ce9cbc8c29ebc1b4559d027df545e6cdd287&end=eaf73cd9c4fc608d616056b232ccb9cbb8df5679&stat=max-rss)
<details>
<summary>Results (primary -1.8%, secondary -3.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.8%  | [2.1%, 3.3%]   | 3     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.9% | [-8.7%, -0.7%] | 13    |
| Improvements ✅ <br /> (secondary) | -3.2% | [-7.1%, -2.0%] | 29    |
| All ❌✅ (primary)                 | -1.8% | [-8.7%, 3.3%]  | 16    |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=c182ce9cbc8c29ebc1b4559d027df545e6cdd287&end=eaf73cd9c4fc608d616056b232ccb9cbb8df5679&stat=cycles:u)
<details>
<summary>Results (primary -1.4%, secondary 2.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.0%  | [2.1%, 5.4%]   | 5     |
| Improvements ✅ <br /> (primary)   | -1.4% | [-1.8%, -1.0%] | 3     |
| Improvements ✅ <br /> (secondary) | -2.3% | [-2.3%, -2.3%] | 1     |
| All ❌✅ (primary)                 | -1.4% | [-1.8%, -1.0%] | 3     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=c182ce9cbc8c29ebc1b4559d027df545e6cdd287&end=eaf73cd9c4fc608d616056b232ccb9cbb8df5679&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.0%, secondary -0.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5%  | [0.0%, 3.9%]   | 38    |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.3% | [-1.8%, -0.1%] | 50    |
| Improvements ✅ <br /> (secondary) | -0.2% | [-2.0%, -0.0%] | 83    |
| All ❌✅ (primary)                 | 0.0%  | [-1.8%, 3.9%]  | 88    |
</details>

**Bootstrap**: 785.339s -> 785.217s (-0.02%)
**Artifact size**: 348.32 MiB -> 348.38 MiB (0.02%)

scottmcm committed:
Save another BB by using `SubUnchecked` instead of a call to `arith_offset`

Probably reasonable anyway since it more obviously drops provenance.

scottmcm committed:
Go back to `Some` instead of transmuting to it.

This adds a few more statements to `next`, but optimizes better in the loops (saving 2 blocks in `forward_loop`, for example)

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 4ddbbbdb50a8dea8f0520591658b5620214ed7f0 with merge d786e76b71a6ff16343a9251ccb9db66784dddda...
<!-- homu: {"type":"TryBuildStarted","head_sha":"4ddbbbdb50a8dea8f0520591658b5620214ed7f0","merge_sha":"d786e76b71a6ff16343a9251ccb9db66784dddda"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13324308184/job/37218695482)
Build commit: d786e76b71a6ff16343a9251ccb9db66784dddda (`d786e76b71a6ff16343a9251ccb9db66784dddda`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13324308184/job/37218695482"},"merge_sha":"d786e76b71a6ff16343a9251ccb9db66784dddda"} -->

rust-timer commented:
Queued d786e76b71a6ff16343a9251ccb9db66784dddda with parent 6dfeab5c9e8a17a6636c1479037baabc4b1e9562, future [comparison URL](https://perf.rust-lang.org/compare.html?start=6dfeab5c9e8a17a6636c1479037baabc4b1e9562&end=d786e76b71a6ff16343a9251ccb9db66784dddda).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (d786e76b71a6ff16343a9251ccb9db66784dddda): [comparison URL](https://perf.rust-lang.org/compare.html?start=6dfeab5c9e8a17a6636c1479037baabc4b1e9562&end=d786e76b71a6ff16343a9251ccb9db66784dddda&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=6dfeab5c9e8a17a6636c1479037baabc4b1e9562&end=d786e76b71a6ff16343a9251ccb9db66784dddda&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5%  | [0.5%, 0.5%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 9.2%  | [9.2%, 9.2%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-2.2%, -0.2%] | 204   |
| Improvements ✅ <br /> (secondary) | -0.5% | [-2.3%, -0.1%] | 101   |
| All ❌✅ (primary)                 | -0.6% | [-2.2%, 0.5%]  | 206   |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=6dfeab5c9e8a17a6636c1479037baabc4b1e9562&end=d786e76b71a6ff16343a9251ccb9db66784dddda&stat=max-rss)
<details>
<summary>Results (primary -1.7%, secondary 0.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range           | count |
|:----------------------------------:|:-----:|:---------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 6.4%  | [2.9%, 9.8%]    | 2     |
| Regressions ❌ <br /> (secondary)  | 3.8%  | [3.4%, 4.2%]    | 2     |
| Improvements ✅ <br /> (primary)   | -3.5% | [-10.0%, -0.8%] | 9     |
| Improvements ✅ <br /> (secondary) | -5.0% | [-5.0%, -5.0%]  | 1     |
| All ❌✅ (primary)                 | -1.7% | [-10.0%, 9.8%]  | 11    |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=6dfeab5c9e8a17a6636c1479037baabc4b1e9562&end=d786e76b71a6ff16343a9251ccb9db66784dddda&stat=cycles:u)
<details>
<summary>Results (primary -1.2%, secondary 1.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 4.5%  | [4.5%, 4.5%]   | 1     |
| Improvements ✅ <br /> (primary)   | -1.2% | [-2.0%, -0.8%] | 11    |
| Improvements ✅ <br /> (secondary) | -2.3% | [-2.3%, -2.3%] | 1     |
| All ❌✅ (primary)                 | -1.2% | [-2.0%, -0.8%] | 11    |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=6dfeab5c9e8a17a6636c1479037baabc4b1e9562&end=d786e76b71a6ff16343a9251ccb9db66784dddda&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.2%, secondary -0.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.0%, 0.8%]   | 31    |
| Regressions ❌ <br /> (secondary)  | 0.0%  | [0.0%, 0.0%]   | 34    |
| Improvements ✅ <br /> (primary)   | -0.4% | [-1.8%, -0.1%] | 61    |
| Improvements ✅ <br /> (secondary) | -0.3% | [-2.0%, -0.0%] | 11    |
| All ❌✅ (primary)                 | -0.2% | [-1.8%, 0.8%]  | 92    |
</details>

**Bootstrap**: 789.679s -> 787.912s (-0.22%)
**Artifact size**: 347.73 MiB -> 348.31 MiB (0.17%)

the8472 commented on library/core/src/slice/iter/macros.rs:
```diff
@@ -154,16 +154,26 @@ macro_rules! iterator {
-                    if is_empty!(self) {
-                        None
+                    if T::IS_ZST {
+                        let len = end_or_len.addr();
```
Would an as cast be more lightweight here? With ZSTs there should be no questions about provenance.

the8472 commented on library/core/src/slice/iter/macros.rs:
```diff
@@ -154,16 +154,26 @@ macro_rules! iterator {
-                // safe since we check if the iterator is empty first.
+                let ptr = self.ptr;
+                let end_or_len = self.end_or_len;
+                // SAFETY: Type invariants.
```
The safety comment is a bit too sloppy imo. At least it should say something like "same as above" if you want to avoid repetition. Or maybe split it into two unsafe blocks, one for each arm.

scottmcm committed:
Add real safety comments

joboet commented on library/core/src/slice/iter/macros.rs:
```diff
@@ -154,16 +154,39 @@ macro_rules! iterator {
+                    // SAFETY: Now that we know it wasn't empty and we've moved past
+                    // the first one (to avoid giving a duplicate `&mut` next time),
+                    // we can give out a reference to it.
+                    Some({ptr}.$into_ref())
```
Silly question, but why does this use braces?

joboet replied:
Ah, I see. That's clever!

joboet reviewed (approved):
This looks great, and the benchmarks speak for themselves – I think we can justify the one regression.

I have one silly nit, which you are welcome to ignore. Otherwise, r=me

bors commented:
:pushpin: Commit 7add35831955f8fc5ba08dcdb93921cd8de17841 has been approved by `joboet`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=joboet 7add35831955f8fc5ba08dcdb93921cd8de17841 -->
<!-- homu: {"type":"Approved","sha":"7add35831955f8fc5ba08dcdb93921cd8de17841","approver":"joboet","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

rust-timer commented:
Finished benchmarking commit (f04bbc60f8c353ee5ba0677bc583ac4a88b2c180): [comparison URL](https://perf.rust-lang.org/compare.html?start=28b83ee59698ae069f5355b8e03f976406f410f5&end=f04bbc60f8c353ee5ba0677bc583ac4a88b2c180&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - please read the text below


Our benchmarks found a performance regression caused by this PR.
This might be an actual regression, but it can also be just noise.

**Next Steps**:

- If the regression was expected or you think it can be justified,
please write a comment with sufficient written justification, and add
`@rustbot label: +perf-regression-triaged` to it, to mark the regression as triaged.
- If you think that you know of a way to resolve the regression, try to create
a new PR with a fix for the regression.
- If you do not understand the regression or you think that it is just noise,
you can ask the `@rust-lang/wg-compiler-performance` working group for help (members of this group
were already notified of this PR).

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=28b83ee59698ae069f5355b8e03f976406f410f5&end=f04bbc60f8c353ee5ba0677bc583ac4a88b2c180&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.4%  | [0.4%, 0.4%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 8.8%  | [8.8%, 8.8%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.4% | [-1.4%, -0.1%] | 123   |
| Improvements ✅ <br /> (secondary) | -0.5% | [-2.3%, -0.1%] | 70    |
| All ❌✅ (primary)                 | -0.4% | [-1.4%, 0.4%]  | 125   |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=28b83ee59698ae069f5355b8e03f976406f410f5&end=f04bbc60f8c353ee5ba0677bc583ac4a88b2c180&stat=max-rss)
<details>
<summary>Results (primary -2.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 7.3%  | [5.5%, 9.0%]   | 2     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -5.8% | [-9.1%, -1.8%] | 5     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -2.1% | [-9.1%, 9.0%]  | 7     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=28b83ee59698ae069f5355b8e03f976406f410f5&end=f04bbc60f8c353ee5ba0677bc583ac4a88b2c180&stat=cycles:u)
<details>
<summary>Results (primary -1.2%, secondary 0.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.8%  | [3.8%, 3.8%]   | 1     |
| Improvements ✅ <br /> (primary)   | -1.2% | [-2.3%, -0.8%] | 8     |
| Improvements ✅ <br /> (secondary) | -2.0% | [-2.0%, -2.0%] | 1     |
| All ❌✅ (primary)                 | -1.2% | [-2.3%, -0.8%] | 8     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=28b83ee59698ae069f5355b8e03f976406f410f5&end=f04bbc60f8c353ee5ba0677bc583ac4a88b2c180&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.1%, secondary -0.3%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3%  | [0.0%, 0.8%]   | 10    |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.2% | [-0.4%, -0.0%] | 58    |
| Improvements ✅ <br /> (secondary) | -0.3% | [-1.6%, -0.0%] | 11    |
| All ❌✅ (primary)                 | -0.1% | [-0.4%, 0.8%]  | 68    |
</details>

**Bootstrap**: 774.026s -> 774.477s (0.06%)
**Artifact size**: 360.27 MiB -> 361.04 MiB (0.21%)

rylev commented:
Perf improvements vastly outweigh the regressions

@rustbot label: +perf-regression-triaged
